relay-profiling/src/sample.rs

```diff
@@ -0,0 +1,429 @@
+            // https://github.com/microsoft/plcrashreporter/blob/748087386cfc517936315c107f722b146b0ad1ab/Source/PLCrashAsyncThread_arm.c#L84
+            (Platform::Cocoa, "arm64") | (Platform::Cocoa, "arm64e") => 0x0000000FFFFFFFFF,
+            _ => return,
+        };
+        for frame in &mut self.frames {
+            frame.strip_pointer_authentication_code(addr);
+        }
+    }
+}
+
+#[derive(Default, Debug, Serialize, Deserialize, Clone)]
+struct DebugMeta {
+    images: Vec<NativeDebugImage>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct OSMetadata {
+    name: String,
+    version: String,
+
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    build_number: Option<String>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct RuntimeMetadata {
+    name: String,
+    version: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct DeviceMetadata {
+    architecture: String,
+
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    is_emulator: Option<bool>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    locale: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    manufacturer: Option<String>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    model: Option<String>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+enum Version {
+    #[serde(rename = "1")]
+    V1,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct SampleProfile {
+    version: Version,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    debug_meta: Option<DebugMeta>,
+
+    device: DeviceMetadata,
+    os: OSMetadata,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    runtime: Option<RuntimeMetadata>,
+
+    #[serde(default, skip_serializing_if = "String::is_empty")]
+    environment: String,
+    #[serde(alias = "profile_id")]
+    event_id: EventId,
+    platform: Platform,
+    profile: Profile,
+    release: String,
+    timestamp: DateTime<Utc>,
+
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    transactions: Vec<TransactionMetadata>,
+}
+
+impl SampleProfile {
+    fn valid(&self) -> bool {
+        match self.platform {
+            Platform::Cocoa => {
+                self.os.build_number.is_some()
+                    && self.device.is_emulator.is_some()
+                    && self.device.locale.is_some()
+                    && self.device.manufacturer.is_some()
+                    && self.device.model.is_some()
+            }
+            Platform::Python => self.runtime.is_some(),
+            Platform::Node => self.runtime.is_some(),
+            _ => true,
+        }
+    }
+
+    /// Removes a sample when it's the only sample on its thread
+    fn remove_single_samples_per_thread(&mut self) {
+        let mut sample_count_by_thread_id: HashMap<u64, u32> = HashMap::new();
+
+        for sample in self.profile.samples.iter() {
```

Comment:
```suggestion
        for sample in self.profile.samples {
```

you shouldn't have to use `.iter()` explicitly here

Reply:
I removed it but I needed to borrow.